quickwit-metastore/src/checkpoint.rs

```diff
@@ -0,0 +1,461 @@
+    }
+
+    fn check_compatibility(&self, delta: &CheckpointDelta) -> Result<(), IncompatibleCheckpoint> {
+        info!(delta=?delta, checkpoint=?self);
+        for (delta_partition, delta_position) in &delta.per_partition {
+            let position = if let Some(position) = self.per_partition.get(delta_partition) {
+                position
+            } else {
+                continue;
+            };
+            match position.cmp(&delta_position.from) {
+                Ordering::Equal => {}
+                Ordering::Less => {
+                    warn!(cur_pos=?position, delta_pos_from=?delta_position.from,partition=?delta_partition, "Some positions were skipped.");
+                }
+                Ordering::Greater => {
+                    return Err(IncompatibleCheckpoint {
+                        partition_id: delta_partition.clone(),
+                        current_position: position.clone(),
+                        delta_position_from: delta_position.from.clone(),
+                    });
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Try and apply a delta.
+    ///
+    /// We accept a delta as long as it comes after the current checkpoint,
+    /// for all partitions.
+    ///
+    /// We accept a delta that is not perfected chained after a checkpoint,
+    /// as gaps may happen. For instance, assuming a Kafka source, if the indexing
+    /// pipeline is down for more than the retention period.
+    ///
+    ///    |    Checkpoint & Delta        | Outcome                     |
+    ///    |------------------------------|-----------------------------|
+    ///    | (?..a] (b..c] with a = b     | Compatible                  |
+    ///    |  (..a] (b..c] with b > a     | Compatible                  |
+    ///    |  (..a] (b..c]  with b < a    | Incompatible                |
+    ///
+    /// If the delta is compatible, returns an error without modifying the original checkpoint.
+    pub fn try_apply_delta(
+        &mut self,
+        delta: CheckpointDelta,
+    ) -> Result<(), IncompatibleCheckpoint> {
+        self.check_compatibility(&delta)?;
+        for (partition_id, partition_position) in delta.per_partition {
+            self.per_partition
+                .insert(partition_id, partition_position.to);
+        }
+        Ok(())
+    }
+}
+
+impl fmt::Debug for Checkpoint {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("Ckpt(")?;
+        for (i, (partition_id, position)) in self.per_partition.iter().enumerate() {
+            f.write_str(&partition_id.0)?;
+            f.write_str(":")?;
+            f.write_str(&position.0)?;
+            let is_last = i == self.per_partition.len() - 1;
+            if !is_last {
+                f.write_str(" ")?;
+            }
+        }
+        f.write_str(")")?;
+        Ok(())
+    }
+}
+
+/// A partition delta represents an interval (from, to] over a partition of a source.
```

Comment:
how do we cover the first element in a source with (from, to]?

Reply:
`Position::default()`. This is the empty string.